File: Client/game/respawnScreen/respawnScreen.cpp
#include "respawnScreen.hpp"

void RespawnScreen::init() {
    back_rect.orientation = gfx::CENTER;
    back_rect.fill_color.a = TRANSPARENCY;
    back_rect.blur_radius = BLUR;
    back_rect.shadow_intensity = SHADOW_INTENSITY;
    back_rect.border_color = BORDER_COLOR;
    back_rect.smooth_factor = 3;
    
    you_died_text.setScale(4);
    you_died_text.orientation = gfx::CENTER;
    
    respawn_button.setScale(3);
    respawn_button.orientation = gfx::CENTER;
}

void RespawnScreen::loadTextures() {
    you_died_text.loadFromSurface(gfx::textToSurface("You Died"));
    respawn_button.loadFromSurface(gfx::textToSurface("Respawn"));
    
    back_rect.w = respawn_button.getWidth() + 100;
    back_rect.y = -2 * back_rect.h;
}

bool RespawnScreen::onKeyUp(gfx::Key key) {
    if(key == gfx::Key::MOUSE_LEFT && respawn_button.isHovered(getMouseX(), getMouseY())) {
        Packet packet;
        packet << ClientPacketType::PLAYER_RESPAWN;
        networking->sendPacket(packet);
        return true;
    }
    return false;
}

void RespawnScreen::render() {
    if(players->getMainPlayer())
        first_time = false;
    is_active = !players->getMainPlayer() && !first_time;

    back_rect.h = gfx::getWindowHeight();
    back_rect.y = is_active ? 0 : -2 * back_rect.h;
    you_died_text.y = back_rect.y - you_died_text.getHeight() / 2 - SPACING / 2;
    respawn_button.y = back_rect.y + respawn_button.getHeight() / 2 + SPACING / 2;
    
    float progress = float(back_rect.y + 2 * back_rect.h) / back_rect.h / 2;
    
    if(progress || is_active) {
        gfx::RectShape(0, 0, gfx::getWindowWidth(), gfx::getWindowHeight()).render({200, 0, 0, (unsigned char)(100 * progress)});
        back_rect.render();
        you_died_text.render();
        respawn_button.render(getMouseX(), getMouseY(), getKeyState(gfx::Key::MOUSE_LEFT));
    }
}

File: client/game/respawn_screen.rs
use crate::client::game::networking::ClientNetworking;
use crate::libraries::events::Event;
use crate::libraries::graphics as gfx;
use crate::libraries::ui;
use crate::libraries::ui::{BaseUiElement, UiElement};
use crate::shared::packet::Packet;
use crate::shared::players::RespawnPacket;
use anyhow::Result;

use crate::libraries::ui::UiContext;
pub struct RespawnScreen {
    respawn_button: ui::Button,
    respawn_text: ui::Sprite,
    back_rect: ui::RenderRect,
    pub is_shown: bool,
}

impl RespawnScreen {
    pub fn new() -> Self {
        Self {
            respawn_button: ui::Button::new(|| {}),
            respawn_text: ui::Sprite::new(),
            back_rect: ui::RenderRect::new(gfx::FloatPos(0.0, 0.0), gfx::FloatSize(0.0, 0.0)),
            is_shown: false,
        }
    }

    pub fn init(&mut self, graphics: &gfx::GraphicsContext) {
        self.respawn_button.orientation = ui::CENTER;
        self.respawn_button.texture = gfx::Texture::load_from_surface(&graphics.font.create_text_surface("Respawn", None));
        self.respawn_button.scale = 3.0;

        self.respawn_text.set_texture(gfx::Texture::load_from_surface(&graphics.font.create_text_surface("You died", None)));
        self.respawn_text.pos.1 = -100.0;
        self.respawn_text.orientation = ui::CENTER;
        self.respawn_text.scale = 3.0;

        self.back_rect.orientation = ui::CENTER;
        self.back_rect.size.0 = 300.0;
        self.back_rect.fill_color = ui::BLACK;
        self.back_rect.fill_color.a = ui::TRANSPARENCY;
        self.back_rect.border_color = ui::BORDER_COLOR;
        self.back_rect.blur_radius = ui::BLUR;
        self.back_rect.shadow_intensity = ui::SHADOW_INTENSITY;
    }

    pub fn render(&mut self, graphics: &mut gfx::GraphicsContext) {
        if self.is_shown {
            gfx::Rect::new(gfx::FloatPos(0.0, 0.0), graphics.get_window_size()).render(graphics, gfx::Color::new(255, 0, 0, 100));

            let window_container = ui::Container::default(graphics);

            self.back_rect.size.1 = graphics.get_window_size().1;
            self.back_rect.render(graphics, &self.back_rect.get_container(graphics, &window_container));

            self.respawn_text.render(graphics, &self.back_rect.get_container(graphics, &window_container));

            self.respawn_button.render(graphics, &self.back_rect.get_container(graphics, &window_container));
            //TODO UI element
        }
    }

    pub fn on_event(&self, event: &Event, graphics: &gfx::GraphicsContext, networking: &mut ClientNetworking) -> Result<()> {
        if self.is_shown {
            if let Some(gfx::Event::KeyPress(gfx::Key::MouseLeft, ..)) = event.downcast::<gfx::Event>() {
                if self.respawn_button.is_hovered(graphics, &self.back_rect.get_container(graphics, &ui::Container::default(graphics))) {
                    let packet = Packet::new(RespawnPacket {})?;
                    networking.send_packet(packet)?;
                }
            }
        }

        Ok(())
    }
}
